### PR TITLE
Add modulePorts and fullModulePorts in DataMirror for Instance

### PR DESCRIPTION
**Pull request by debs-sifive**
### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- Feature (or new API)

#### Desired Merge Strategy

- Squash: The PR will be squashed and merged (choose this if you have no preference).

#### Release Notes

Added `modulePorts` and `fullModulePorts` methods in `DataMirror` that returns all ports on an `Instance` of a module.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels? (Select the most appropriate one based on the "Type of Improvement")
- [ ] Did you mark the proper milestone (Bug fix: `3.6.x`, `5.x`, or `6.x` depending on impact, API modification or big change: `7.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

File: src/test/scala/chiselTests/reflect/DataMirrorSpec.scala [hunk 1]
```diff
@@ -265,4 +265,40 @@ class DataMirrorSpec extends ChiselFlatSpec {
...
+      val instB = (Module(new AddOne)).toInstance
+      val portsB = DataMirror.fullModulePorts(instB)
+    }
+
+    ChiselStage.emitCHIRRTL(new Module {
+      val foo = Module(new Foo)
+      (foo.portsA.map(_._2) should contain).allOf(foo.instA.in, foo.instA.out)
+      (foo.portsB.map(_._2) should contain).allOf(foo.instB.in, foo.instB.out)
+    })
+  }
```
Comment (jackkoenig) [inline src/test/scala/chiselTests/reflect/DataMirrorSpec.scala:303]: Generally looks good, but please use a Proto module type that has ports that illustrates the difference between `.modulePorts` and `.fullModulePorts` as for `AddOne` they give the same answer 🙂

Comment (debs-sifive) [reply src/test/scala/chiselTests/reflect/DataMirrorSpec.scala:300]: oops yes, can you check the new test?

File: src/test/scala/chiselTests/reflect/DataMirrorSpec.scala [hunk 2]
```diff
@@ -265,4 +265,37 @@ class DataMirrorSpec extends ChiselFlatSpec {
...
+
+  "modulePorts and fullModulePorts" should "return an Instance of a module's IOs" in {
+    import chiselTests.experimental.hierarchy.Examples.AddOne
+    @instantiable
+    class Bar extends Module {
+      @public val io = IO(new Bundle {
+        val vec = Vec(2, Bool())
+        val x = UInt(4.W)
+      })
+      dontTouch(io)
```
Comment (jackkoenig) [inline src/test/scala/chiselTests/reflect/DataMirrorSpec.scala:277]: No need for dontTouch as we are only emitting CHIRRTL anyway.

File: src/test/scala/chiselTests/reflect/DataMirrorSpec.scala [hunk 3]
```diff
@@ -265,4 +265,37 @@ class DataMirrorSpec extends ChiselFlatSpec {
       foo.ports1 should be(Seq(foo.in, foo.out, foo.other))
     })
   }
+
+  "modulePorts and fullModulePorts" should "return an Instance of a module's IOs" in {
+    import chiselTests.experimental.hierarchy.Examples.AddOne
```
Comment (jackkoenig) [inline src/test/scala/chiselTests/reflect/DataMirrorSpec.scala:270]: I believe this import is now unused
```suggestion
```

File: src/test/scala/chiselTests/reflect/DataMirrorSpec.scala (hunk 2 above)
Comment (jackkoenig) [reply src/test/scala/chiselTests/reflect/DataMirrorSpec.scala:277]:
```suggestion
```

File: src/test/scala/chiselTests/reflect/DataMirrorSpec.scala [hunk 4]
```diff
@@ -265,4 +265,37 @@ class DataMirrorSpec extends ChiselFlatSpec {
...
+      val foo = Module(new Foo)
+      foo.portsA.map(_._1) should contain("io")
+      foo.portsA.map(_._1) should not contain ("io_x")
+      (foo.portsB.map(_._1) should contain).allOf(
+        "io",
+        "io_x",
+        "io_vec",
+        "io_vec_0",
+        "io_vec_1"
+      )
```
Comment (jackkoenig) [inline src/test/scala/chiselTests/reflect/DataMirrorSpec.scala:298]: Can we provide a tighter check here?
```suggestion
      foo.portsA.map(_._1) should be(Seq("io"))
      (foo.portsB.map(_._1) should be(
        Seq(
          "io",
          "io_x",
          "io_vec",
          "io_vec_0",
          "io_vec_1"
        )
      )
```

**Review by jackkoenig — APPROVED** [review]
Looks good but please see comments on tests.

Also this is failing to generate ScalaDoc due to ambiguous links. Theoretically it's possible to disambiguate them but I find ScalaDoc to be super flakey on that so just replace the `[[` and `]]` around `[[DataMirror.modulePorts]]` with `backticks` and call it a day. You can reproduce the failure with `sbt "unipublish / unidoc"`

**Pull request merged**